packages/server-wallet/src/protocols/ledger-funding.ts

```diff
@@ -0,0 +1,158 @@
+import _ from 'lodash';
+import {
+  allocateToTarget,
+  checkThat,
+  isSimpleAllocation,
+  SimpleAllocation,
+} from '@statechannels/wallet-core';
+
+import {Bytes32} from '../type-aliases';
+
+import {Protocol, ProtocolResult, ChannelState} from './state';
+import {
+  LedgerProtocolAction,
+  MarkLedgerFundingRequestsAsComplete,
+  noAction,
+  SignLedgerStateForRequests,
+  signState,
+} from './actions';
+
+export type ProtocolState = {
+  fundingChannel: ChannelState;
+  channelsPendingRequest: ChannelState[];
+  channelsWithInflightRequest: ChannelState[];
+};
+
+const newOutcomeBasedOnMyPendingUpdates = (
+  supportedOutcome: SimpleAllocation,
+  channelsPendingRequest: ChannelState[]
+): SimpleAllocation =>
+  // This could fail at some point if there is no longer space to fund stuff
+  // TODO: Handle that case
+  // } catch (e) {
+  //   if (e.toString() === 'Insufficient funds in fundingChannel channel')
+  // }
+  // TODO: All this usage of checkThat is annoying
+  channelsPendingRequest.reduce(
+    (outcome, {channelId, supported}) =>
+      checkThat(
+        allocateToTarget(
+          outcome,
+          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+          checkThat(supported!.outcome, isSimpleAllocation).allocationItems,
+          channelId
+        ),
+        isSimpleAllocation
+      ),
+    supportedOutcome
+  );
+
+const outcomeMergedWithLatestState = (
+  latestOutcome: SimpleAllocation,
+  newOutcome: SimpleAllocation
+): SimpleAllocation => ({
+  type: 'SimpleAllocation',
+  assetHolderAddress: newOutcome.assetHolderAddress,
+  allocationItems: _.intersection(latestOutcome.allocationItems, newOutcome.allocationItems),
+});
+
+const computeNewOutcome = ({
+  fundingChannel: {supported, latest, latestSignedByMe, channelId},
+  channelsPendingRequest,
+}: ProtocolState): SignLedgerStateForRequests | false => {
+  if (!supported) return false;
+  if (!latestSignedByMe) return false;
+
+  const supportedOutcome = checkThat(supported.outcome, isSimpleAllocation);
+  const latestOutcome = checkThat(latest.outcome, isSimpleAllocation);
+  const myLatestOutcome = checkThat(latestSignedByMe.outcome, isSimpleAllocation);
+
+  const counterPartyProposedNewUpdate = latest.turnNum === latestSignedByMe.turnNum + 2;
```

Comment:
Is this assuming a 2 party channel?

Reply:
Yeah, it could be `+ n` I suppose... need to think more about how that works with CRDTs in this case.